ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/CommandService.scala

```diff
@@ -0,0 +1,221 @@
+    for {
+      alice <- allocateParty()
+      transactionId <- session.bindings.submitAndWaitForTransactionId(
+        alice,
+        context.applicationId,
+        context.nextCommandId(),
+        Seq(Dummy(alice).create.command.command))
+    } yield {
+      assert(
+        transactionId.nonEmpty,
+        s"The returned transaction ID should be a non empty string, but was '$transactionId'.")
+    }
+  }
+
+  private val submitAndWaitForTransaction = LedgerTest(
+    "CSsubmitAndWaitForTransaction",
+    "SubmitAndWaitForTransaction returns a transaction") { implicit context =>
+    for {
+      alice <- allocateParty()
+      transaction <- session.bindings.submitAndWaitForTransaction(
+        alice,
+        context.applicationId,
+        context.nextCommandId(),
+        Seq(Dummy(alice).create.command.command))
+    } yield {
+      assert(
+        transaction.transactionId.nonEmpty,
+        s"The returned transaction ID should be a non empty string, but was '${transaction.transactionId}'.")
+
+      assert(
+        transaction.events.size == 1,
+        s"The returned transaction should contain 1 event, but contained ${transaction.events.size}")
+
+      val event = transaction.events.head
+      assert(
+        event.event.isCreated,
+        s"The returned transaction should contain a created-event, but was ${event.event}"
+      )
+
+      assert(
+        event.getCreated.getTemplateId == Dummy.id.unwrap,
+        s"The template ID of the created-event should by ${Dummy.id}, but was ${event.getCreated.getTemplateId}"
+      )
+    }
+  }
+
+  private val submitAndWaitForTransactionTree = LedgerTest(
+    "CSsubmitAndWaitForTransactionTree",
+    "SubmitAndWaitForTransactionTree returns a transaction tree") { implicit context =>
+    for {
+      alice <- allocateParty()
+      transactionTree <- session.bindings.submitAndWaitForTransactionTree(
+        alice,
+        context.applicationId,
+        context.nextCommandId(),
+        Seq(Dummy(alice).create.command.command))
+    } yield {
+      assert(
+        transactionTree.transactionId.nonEmpty,
+        s"The returned transaction ID should be a non empty string, but was '${transactionTree.transactionId}'.")
+
+      assert(
+        transactionTree.eventsById.size == 1,
+        s"The returned transaction tree should contain 1 event, but contained ${transactionTree.eventsById.size}")
+
+      val event = transactionTree.eventsById.head._2
+      assert(
+        event.kind.isCreated,
+        s"The returned transaction tree should contain a created-event, but was ${event.kind}")
+
+      assert(
+        event.getCreated.getTemplateId == Dummy.id.unwrap,
+        s"The template ID of the created-event should by ${Dummy.id}, but was ${event.getCreated.getTemplateId}"
+      )
+    }
+  }
+
+  private val resendingSubmitAndWait = LedgerTest(
+    "CSduplicateSubmitAndWait",
+    "SubmitAndWait returns Empty for a duplicate submission") { implicit context =>
```

Comment:
Could this be better phrased as idempotent with regards to the command identifier?

Reply:
Note: for this to be implemented the suggested `suggestAndWait` method on the `LedgerContext` should allow to optionally pass an arbitrary command identifier (not using `nextCommandId()` can actually serves us best here as we could specify a command identifier that can make it easier to trace the source of the error in a failing test).